components/support/migration/src/main/java/mozilla/components/support/migration/AddonMigration.kt

```diff
@@ -0,0 +1,137 @@
+     */
+    sealed class Failure : AddonMigrationResult() {
+        /**
+         * Failed to query installed add-ons via the [Engine].
+         */
+        internal data class FailedToQueryInstalledAddons(val throwable: Throwable) : Failure() {
+            override fun toString(): String {
+                return "Failed to query installed add-ons: ${throwable::class}"
+            }
+        }
+
+        /**
+         * Failed to migrate some or all of the installed add-ons.
+         */
+        internal data class FailedToMigrateAddons(
+            val migratedAddons: List<WebExtension>,
+            val failedAddons: Map<WebExtension, Exception>
+        ) : Failure() {
+            override fun toString(): String {
+                return "Failed to migrate ${failedAddons.size} of ${failedAddons.size + migratedAddons.size} add-ons."
+            }
+        }
+    }
+}
+
+/**
+ * Helper for migrating add-ons.
+ */
+internal object AddonMigration {
+
+    /**
+     * Performs a migration of all installed add-ons. The only migration step involved
+     * is to make sure we disable all currently unsupported add-ons.
+     */
+    @SuppressWarnings("TooGenericExceptionCaught")
+    suspend fun migrate(engine: Engine): Result<AddonMigrationResult> {
+        val installedAddons =
+        try {
+            getInstalledAddons(engine)
+        } catch (e: Exception) {
+            return Result.Failure(AddonMigrationException(AddonMigrationResult.Failure.FailedToQueryInstalledAddons(e)))
+        }
+
+        if (installedAddons.isEmpty()) {
+            return Result.Success(AddonMigrationResult.Success.NoAddons)
+        }
+
+        val migratedAddons = mutableListOf<WebExtension>()
+        val failures = mutableMapOf<WebExtension, Exception>()
+        installedAddons.forEach { addon ->
+            try {
+                migratedAddons += migrateAddon(engine, addon)
+            } catch (e: Exception) {
+                failures[addon] = e
+            }
+        }
+
+        return if (failures.isEmpty()) {
+            Result.Success(AddonMigrationResult.Success.AddonsMigrated(migratedAddons))
+        } else {
+            val failure = AddonMigrationResult.Failure.FailedToMigrateAddons(migratedAddons, failures)
+            Result.Failure(AddonMigrationException(failure))
+        }
+    }
+
+    private suspend fun getInstalledAddons(engine: Engine): List<WebExtension> {
+        return CompletableDeferred<List<WebExtension>>().also { result ->
+            withContext(Dispatchers.Main) {
+                engine.listInstalledWebExtensions(
+                    onSuccess = { result.complete(it) },
+                    onError = { result.completeExceptionally(it) }
+                )
+            }
+        }.await()
+    }
+
+    private suspend fun migrateAddon(engine: Engine, addon: WebExtension): WebExtension {
+        // TODO (For final migration): check if addon is supported (against AMO via
```

Comment:
Can you file an issue for this and link it here? Don't want to loose track of this.

Reply:
Done, linked to the issue now.